### PR TITLE
Fix best model

### PR DESCRIPTION
**Pull request by wli51**
Fixing incorrect best model updating and saving behavior, and related refactor of early termination functionality to reduce trainer module size. 

Note: `src/virtual_stain_flow/trainers/trainer_utils/early_stop.py` is mostly migrated tests from `tests/trainers/test_abstract_trainer.py` refactored to be testable independent of the trainer realization.

### CODE REVIEW

File: src/virtual_stain_flow/trainers/AbstractTrainer.py [hunk 1]
```diff
@@ -124,6 +118,15 @@ def _init_state(
...
         self._val_metrics = defaultdict(list)
 
+        # Early stopping state
+        self._early_stop_helper = EarlyStopHelper(
+            best_mode=early_termination_mode,
+            trainer_val_losses_ref=self._val_losses,
+            trainer_val_metrics_ref=self._val_metrics,
+            best_metric_name=early_termination_metric,
+            enabled=bool(self._val_loader),
+        )
```
Comment (MattsonCam) [inline src/virtual_stain_flow/trainers/AbstractTrainer.py:141]: Up to you, but it may be worth mentioning somewhere in the docs that  with `best_metric_name=None` and with a validation loader present, it falls back to the first validation loss and populates `best_model` anyway

File: src/virtual_stain_flow/models/model.py [hunk 2]
```diff
@@ -148,3 +158,18 @@ def out_channels(self) -> int:
     def out_activation(self) -> torch.nn.Module:
         """Activation function for the output layer."""
         return self._out_activation
+
+
+def clone_model(model: torch.nn.Module) -> torch.nn.Module:
+    try:
+        snapshot = copy.deepcopy(model)
```
Comment (MattsonCam) [inline src/virtual_stain_flow/models/model.py:165]: I think this can double model memory usage on GPU

File: src/virtual_stain_flow/trainers/AbstractTrainer.py (hunk 1 above)
Comment (MattsonCam) [inline src/virtual_stain_flow/trainers/AbstractTrainer.py:141]: If `early_termination_metric` is set but no validation loader is present, early stopping is silently disabled here via `enabled=bool(self._val_loader)`. May consider raising an error instead

**Review by MattsonCam — APPROVED** [review]
This LGTM, good job @wli51 !

File: src/virtual_stain_flow/models/model.py (hunk 2 above)
Comment (wli51) [reply src/virtual_stain_flow/models/model.py:165]: Good catch that I didn't notice because UNets and even UNeXts are themselves quite small on VRAM and the deep copy does not also copy back propagation specific things to the occupation of VRAM by deep copy alone can be negligible. However this is a bad pattern and will fail as the model gets bigger. Fixed by only copying model once at start of training during initialization of trainer in CPU, and subsequent best model updates become only through state dict copy

File: src/virtual_stain_flow/trainers/AbstractTrainer.py (hunk 1 above)
Comment (wli51) [reply src/virtual_stain_flow/trainers/AbstractTrainer.py:141]: Now raises error!

Comment (wli51) [reply src/virtual_stain_flow/trainers/AbstractTrainer.py:141]: Added documentation
